src/writer.rs

```diff
@@ -0,0 +1,373 @@
+    let mut writer = BufWriter::new(f);
+    if is_gz_file(&path) {
+        let mut buffer = Vec::new();
+        write(&mut buffer, header, data);
+
+        let mut e = GzEncoder::new(Vec::new(), Compression::default());
+        e.write_all(&buffer).unwrap();
+
+        let buffer = e.finish().unwrap();
+        writer.write_all(&buffer).unwrap();
+    } else {
+        write(&mut writer, header, data);
+    }
+}
+
+fn write<A, S, D, W>(
+    writer: &mut W,
+    header: NiftiHeader,
+    data: &ArrayBase<S, D>,)
+    where S: Data<Elem=A>,
+          A: TypeTool<A> + Copy + Div<Output = A> + FromPrimitive + ScalarOperand + Sub<Output = A>,
+          D: Dimension + RemoveAxis,
+          W: WriteBytesExt
+{
+    writer.write_i32::<B>(header.sizeof_hdr).unwrap();
+    writer.write_all(&header.data_type).unwrap();
+    writer.write_all(&header.db_name).unwrap();
+    writer.write_i32::<B>(header.extents).unwrap();
+    writer.write_i16::<B>(header.session_error).unwrap();
+    writer.write_u8(header.regular).unwrap();
+    writer.write_u8(header.dim_info).unwrap();
+    for s in &header.dim {
+        writer.write_u16::<B>(*s).unwrap();
+    }
+    writer.write_f32::<B>(header.intent_p1).unwrap();
+    writer.write_f32::<B>(header.intent_p2).unwrap();
+    writer.write_f32::<B>(header.intent_p3).unwrap();
+    writer.write_i16::<B>(header.intent_code).unwrap();
+    writer.write_i16::<B>(header.datatype).unwrap();
+    writer.write_i16::<B>(header.bitpix).unwrap();
+    writer.write_i16::<B>(header.slice_start).unwrap();
+    for f in &header.pixdim {
+        writer.write_f32::<B>(*f).unwrap();
+    }
+    writer.write_f32::<B>(header.vox_offset).unwrap();
+    writer.write_f32::<B>(header.scl_slope).unwrap();
+    writer.write_f32::<B>(header.scl_inter).unwrap();
+    writer.write_i16::<B>(header.slice_end).unwrap();
+    writer.write_u8(header.slice_code).unwrap();
+    writer.write_u8(header.xyzt_units).unwrap();
+    writer.write_f32::<B>(header.cal_max).unwrap();
+    writer.write_f32::<B>(header.cal_min).unwrap();
+    writer.write_f32::<B>(header.slice_duration).unwrap();
+    writer.write_f32::<B>(header.toffset).unwrap();
+    writer.write_i32::<B>(header.glmax).unwrap();
+    writer.write_i32::<B>(header.glmin).unwrap();
+
+    writer.write_all(&header.descrip).unwrap();
+    writer.write_all(&header.aux_file).unwrap();
+    writer.write_i16::<B>(header.qform_code).unwrap();
+    writer.write_i16::<B>(header.sform_code).unwrap();
+    for f in &[header.quatern_b, header.quatern_c, header.quatern_d,
+               header.quatern_x, header.quatern_y, header.quatern_z] {
+        writer.write_f32::<B>(*f).unwrap();
+    }
+    for f in header.srow_x.iter().chain(&header.srow_y).chain(&header.srow_z) {
+        writer.write_f32::<B>(*f).unwrap();
+    }
+    writer.write_all(&header.intent_name).unwrap();
+    writer.write_all(&header.magic).unwrap();
+
+    // Empty 4 bytes after the header
+    writer.write_u32::<B>(0).unwrap();
```

Comment:
Eventually, we ought to support writing extension data as well. We can keep this gap documented in the mean time.